### ISSUE TITLE
[RFC] Scopes used within documentation blocks

### ISSUE THREAD

**Issue by rchl** [sublimehq/Packages#2569]
Various languages support documentation blocks that can include semantic tags. This RFC's intention is to try to define conventions for the scopes used for those tags.

## Examples of existing documentation blocks

<details>
<summary>JSDoc (Javascript)</summary>

```js
/**
 * Does something
 *
 * @param {number} foo The param description
 * @return {Promise<bool>} The result description
 */
```

</details>

<details>
<summary>PHPDoc (PHP)</summary>

```php
/**
 * Some text {@link http://foo.com}
 * @access public
 * @param string $argument1 This is the description.
 */
```

</details>

<details>
<summary>Documentation comments (C#)</summary>

```cs
/// <summary>This method changes the point's location to
///    the given coordinates.</summary>
/// <param name="xor">the new x-coordinate.</param>
/// <param name="yor">the new y-coordinate.</param>
```

</details>

<details>
<summary>JavaDoc (Java)</summary>

```java
/**
 * Blabla {@link ImageObserver} object will be notified as the
 * conversion process progresses.
 *
 * @param observer  the image observer to be notified as more
 *                  of the image is converted.  May be
 *                  <code>null</code>
 * @return          <code>true</code> if the image is completely
 *                  loaded and was painted successfully;
 *                  <code>false</code> otherwise.
 * @see             Image
 */
```

</details>

<details>
<summary>perlpod (Perl)</summary>

```perl
=head2 Methods

=over 12

=item C<new>

Returns a new My::Module object.

=item C<as_string>

Returns a stringified representation of
the object. This is mainly for debugging
purposes.

=back
```

</details>

### PHPDoc

Current implementation only scopes tags (including inline) and "storage modifers" (for example `public` in `@access public`).

Used scopes: `keyword.other.phpdoc`, `storage.modifier`, `meta.tag.inline.phpdoc`

<details>
<summary>PHPDoc Example</summary>

        ```
        /**
         * Some text {@link http://foo.com}
         *           ^^^^^^^^^^^^^^^^^^^^^^^^^ meta.tag.inline.phpdoc.php
         *            ^^^^^ keyword.other.phpdoc.php
         *
         * @param string $argument1 This is the description.
        // ^^^^^^ keyword.other.phpdoc.php
         *
         * @access public
        //         ^^^^^^ storage.modifier.php
         */
        ```

</details>

### C#

Used scopes: `entity.name.tag.*`

<details>
<summary>C# Example</summary>

```
    /// <summary>
//*     ^ punctuation.definition.tag.begin
//*      ^^^^^^^ entity.name.tag.begin
//*             ^ punctuation.definition.tag.end
    /// </summary>
//*       ^^^^^^^ entity.name.tag.end
    /// <see href="http://foo.com">Reference</see>
//*      ^^^ entity.name.tag
//*          ^^^^ entity.other.attribute-name
//*              ^ punctuation.separator.argument.value
```

</details>

### JavaDoc

Used scopes: `meta.block-tag.javadoc`, `keyword.other.documentation.param.javadoc`, `variable.parameter.javadoc`, `meta.tag`, `punctuation.section.inline-tag.begin.javadoc`, `keyword.other.documentation.*.javadoc`

<details>
<summary>JavaDoc Example</summary>

```
  /**
   * @param paramName Some description
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ meta.block-tag.javadoc
//   ^^^^^^ keyword.other.documentation.param.javadoc
//          ^^^^^^^^^ variable.parameter.javadoc
   *                  that spans <i>several</i> lines.
//                               ^^^ meta.tag
   * {@code} {@literal}
//   ^ punctuation.section.inline-tag.begin.javadoc
//    ^^^^^ keyword.other.documentation.code-or-literal.javadoc
   */
```

</details>

### perlpod

Used scopes: `entity.name.tag.pod.perl`, `entity.name.tag.bold.perl`

<details>
<summary>perlpod Example</summary>

```
=head1 B<--param>
#^^^^^ entity.name.tag.pod.perl
#      ^^^^^^^^^^ markup.heading.perl
#      ^ entity.name.tag.bold.perl
```

</details>

### JSDoc

JSDoc handling is currently not implemented (attempt at adding initial support is at #2557). The syntax is very similar to JavaDoc but with the notable addition of being able to specify the type of `@param` (and other block tags).

<details>
<summary>JSDoc Example</summary>

```js
   /**
    * @param {Foo.bar} param The param description
    */
```

</details>

## Questions

 - What common scopes can we agree upon to use for those documentation blocks?

The JSDoc/JavaDoc/PHPDoc systems are very similar and I think it would make sense that those are consistent. We can possibly mostly ignore `perl` and `c#` in this discussion as those are quite different and probably can't share much with the other ones.

 - How will those scopes affect syntax highlighting and would there be enough clarify and not too much noise?

One might be concerned about all those tags, types, and param names created a lot of extra noise and distracting from the actual code.

 - Do we need new scopes that don't have associated colors and if so, do we new color bindings in color schemes for those?

I suppose there is a possibility of creating more dimmed/muted colors for those scopes but since that would require adding new color<=>scope bindings in color schemes, it would be a big undertaking and would require all color schemes to be updated.

**Comment by jwortmann** [comment]
Another example from R (roxygen), using `keyword.other` and `variable.parameter`:

```r
#' @param xyz abcde
#^^^^^^^^^^^^^^^^^^ comment.line.roxygen.r
#  ^^^^^^ keyword.other.r
#         ^^^ variable.parameter.r
```

**Comment by deathaxe** [comment]
C# is based on XML and JavaDoc is based on HTML. Scoping XML like tags as `entity.name.tag` is very ok.

Perl uses `entity.name.tag` inspired by C# and JavaDoc as `=pod` and friends have a very similar meaning but with "slightly" different syntax.

Some more popular color schemes address `entity.name.tag` to look different than `keyword` scopes, which makes it easier to distinguish them from normal code without the need to address the whole comment scope as well.

JavaDoc uses both HTML tags and `@<name>` keywords. It even distinguishes between block and inline `{@<name> ...}`. I guess the ladder ones are scoped `keyword.other` to ensure different highlighting than ordinary HTML tags even in simpler color schemes.

Nevertheless inline tags like `{@code whatever}` are a shorthand representation of `<code>whatever</code>`. Thus scoping the `@code` `entity.name.tag.documentation` would be appropriate as well.

Same may apply to block `@<name>` tags. `entity.name.tag` might even be a better choise than `keyword.other` as it might easily be possible the following word to denote a `keyword` or `storage` at some point.

```Java
/**
 * {Description}
 *
 * @param void function() : This is the function
 **/
```

As a result

- `@param`  would use `entity.name.tag.documentation`
- `<tag></tag>` would use `entity.name.tag.[html|xml]`

This would makes those look the same in most color schemes, while still enables us to distinguish them.

**Comment by wbond** [comment]
Looking over the scope naming docs, I think `entity.other.attribute-name.documentation` would be a better fit than `entity.name.tag.documentation` for the form `@param`. We are really defining attributes of the code, as opposed to writing markup.

For the HTML/XML forms, they can definitely use the `entity.name.tag.(html|xml)` scopes.

**Comment by rchl** [comment]
Color-wise, `entity.other.attribute-name.documentation` kinda blends with function definitions and makes it hard to see boundaries of functions IMO. But possibly it's just a matter of getting used to...

I feel like we'll never get to an agreement if we will want to make both the scope and color perfect...

`entity.other.attribute-name.documentation`

![Screenshot 2020-11-12 at 15 28 09](https://user-images.githubusercontent.com/153197/98952378-ad6e2480-24fb-11eb-8e50-9c4c8550a6c2.png)

`entity.name.jsdoc`

![Screenshot 2020-11-12 at 15 28 53](https://user-images.githubusercontent.com/153197/98952495-cecf1080-24fb-11eb-875d-d178584f3d75.png)

**Comment by michaelblyons** [comment]
@rchl You can override your color scheme to make the colors whatever you want. I also have a suspicion that if a standard is reached, then the default color schemes (like Mariana in your screenshot) may add rules to target them. They already have a couple special cases (including one for YAML kvp keys that this repo recently broke).

**Comment by rchl** [comment]
Yes but I think we are here to define good defaults which should include decent color defaults.

I'm not sure if it's an option to change all the schemes now to introduce new color mapping since that seems like a big task in itself.

**Comment by wbond** [comment]
> I'm not sure if it's an option to change all the schemes now to introduce new color mapping since that seems like a big task in itself.

The biggest point is that currently vast swaths of color schemes may already scope `<a>` the same as `export function`. You really can't factor in the color of a token in terms of what scope to pick. There are effectively 10 "basic" colors - red, orange, yellow, green, blue, cyan, purple, pink, brown, white/black (depending on bg color). All of the color schemes in the wild won't necessarily scope `entity.name.tag` (primarily seen by users as HTML tags) a different color than `storage`.

If we feel strongly enough to distinguish documentation attributes/tags from the code most often seen around it, then we will do that in the default syntaxes. People who care about maintaining their syntaxes will similarly make them look "better", especially if we come up with a standard recommendation for scoping.

**Comment by rchl** [comment]
OK then. We can pick the perfect scopes and deal with the rest later, I guess.

Going back to your suggestion... does your suggested `entity.other.attribute-name.documentation` apply to `@param` block tag exclusively? Because there are [dozen of other block tags](https://jsdoc.app/index.html#block-tags) and it would insane to try to pick different scopes for all of them.

And there are also inline tags like `{@link foo}`.

**Comment by wbond** [comment]
I think it should be applied to all `@\w+` tokens. I don't think we really need to scope the `@` as `punctuation.definition`, but some people might like that.

**Comment by jwortmann** [comment]
I think a main argument for the `entity.name.tag.documentation` scope for `@param`, `@see`, etc. is that these things are explicitely called "tags" in [JavaDoc](https://docs.oracle.com/en/java/javase/15/docs/specs/javadoc/doc-comment-spec.html#standard-tags), [JSDoc](https://jsdoc.app/#block-tags), [PHPDoc](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#53-tags) ([Doxygen](https://www.doxygen.nl/manual/commands.html) uses the term "command" instead).
On the other hand I can see why `entity.other.attribute-name.documentation` would fit equally for actual attributes like `@param`, `@private` or `@return`, but there might be more general kewords/tags like `@see` or `@todo` which I wouldn't really call "attributes" of a class or function. Not sure which of the two options "tag" or "attribute-name" I would prefer, though.

From the color/highlighting perspective, you can't really argue for one or the other, even within the default color schemes. In Monokai `entity.name.tag` would coincide with the color for `keyword`, while in Mariana the same is the case with `entity.other.attribute-name`.

<hr>

Another small inconsistency I noticed is the "comment" scope in doc comments which are line comments:
* R currently uses `comment.line.roxygen.r`; could this be unified to `comment.line.documentation.r` or possibly `comment.line.documentation.roxygen.r`, so that color schemes don't need an additional rule only for R?
* C# uses `comment.block.documentation.cs` even for line comments (denoted by `///`), this should be changed to `comment.line.documentation.cs` imo
It will mean that both `comment.block.documentation` and `comment.line.documentation` must be listed if a color scheme wants to target certain tokens in all kind of doc comments, but I don't think this is really an issue.

**Comment by deathaxe** [comment]
I guess `comment.block.documentation` is used because it is suggested by PackageDev and a documentation in one line may semantically feel odd. But I agree, it should use `comment.line.documentation` in C#.

**Comment by rwols** [comment]
If I want to search for all docs with the ST API, I'm not really going to care whether it's a line (comment.line) or a block (comment.block), so I'd say use comment.block.documentation even for single lines. Or even comment.documentation.

**Comment by deathaxe** [comment]
So just changing R (former suggestion) or all syntaxes (ladder suggestion).

**Comment by deathaxe** [comment]
I can't associate `@param`, `@see`, ... with attribute as well, honestly, especially as `@code` in `{@code foo()}`  is really a synonym for `<code>foo()</code>` in JavaDoc, so they are really tags, semantically. But can follow those who want to easily distinguish those from xml/html tags if both are present, easily.

**Comment by jwortmann** [comment]
> 
> 
> If I want to search for all docs with the ST API, I'm not really going to care whether it's a line (comment.line) or a block (comment.block), so I'd say use comment.block.documentation even for single lines. Or even comment.documentation.

`view.match_selector()` supports scope selectors, so you can easily use `"comment.block.documentation | comment.line.documentation"` as a selector if you want to find all kind of doc comments.
Since we have `block` and `line` as a 2nd level scope to distinguish between them, why shouldn't we use it with doc comments too? It would feel weird to have these scopes, but then suggest to violate them just within doc comments.

**Comment by rwols** [comment]
> view.match_selector() supports scope selectors, so you can easily use "comment.block.documentation | comment.line.documentation" as a selector if you want to find all kind of doc comments.

I am aware of that. I just don't see it serving any practical purpose.

> Since we have block and line as a 2nd level scope to distinguish between them, why shouldn't we use it with doc comments too? It would feel weird to have these scopes, but then suggest to violate them just within doc comments.

Fair point. I suppose we should adhere to this.

**Comment by deathaxe** [comment]
In the special case of C# one could argue that a triple slash `///` may be a special notation for documentation comments, which are normally appear as a larger block covering multiple lines. There might be no correct or wrong solution here, depending on the point of view. ... speaking as a non-c#-expert.

**Comment by wbond** [comment]
> I think a main argument for the entity.name.tag.documentation scope for @param, @see, etc. is that these things are explicitely called "tags" in JavaDoc, JSDoc, PHPDoc (Doxygen uses the term "command" instead).

That is true, however they are also different in nature with the common use of `entity.name.tag` which is for HTML and XML tags.

https://www.merriam-webster.com/dictionary/tag:

 - Def 5f: "an element of code in a computer document used especially to control format and layout or to establish a hyperlink" seems to fit HTML/XML tags
 - Def 5a "...marker used for identification or classification" seems to fit `@param` better

Do we want to stick with the literal name being the same, even though they have different uses? I think in describing the nature of `@param`, `@return` these aren't *tags* of the following code, but rather *attributes* of the code.

It does appear that for the sake of consistency they also use such tags for markup/linking purposes also.

----

All that to say, I don't feel extremely strongly one way or the other, just that `entity.name.tag` feels like more of a special case for markup tags, and mostly due to backwards compatibility. I don't think we would ever have picked an `entity.name`  scope for something like HTML tags if we were trying to decide today. Rather probably `entity.other.tag`, or maybe a `markup` scope.

**Comment by deathaxe** [comment]
So `@param` is something like 5g, just to call it ATTAG instead of HASHTAG, `#param` :-D (joking)
